### PR TITLE
PR: Migrate completion plugin to the new API

### LINKED ISSUE #12192: Move Completion plugin to use new API
- [ ] Refactor plugins to use the standard plugin API
- [ ] Add a extension manager for extensions
- [ ] Make sure tests pass
- [ ] Move to new API

### PR DESCRIPTION
**Pull request by andfoy**
## Description of Changes
This PR migrates the completion plugin (former CompletionManager) to the new API. As part of this PR, we also introduce a CompletionProvider interface, from which all completion providers inherit. The new CompletionPlugin will take all the providers (whom are registered via entrypoints), initialize them and manage them.

* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

### Issue(s) Resolved

Fixes #12192

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: @andfoy

### CODE REVIEW

File: spyder/plugins/completion/providers/languageserver/widgets/status.py [hunk 1]
```diff
@@ -113,37 +114,28 @@ def get_icon(self):
...
-    def get_current_editor_language(self):
-        """Get current LSP language."""
-        main = self.plugin.main
-        language = _('Unknown')
-
-        if main and main.editor:
-            codeeditor = main.editor.get_current_editor()
-            if codeeditor is not None:
-                language = codeeditor.language
-        return language
```
Comment (ccordoba12) [inline spyder/plugins/completion/providers/languageserver/widgets/status.py:149]: I added this to handle the case where the language server is still not up and running, but you change to another file in the editor and come back here.

How are handling that now?

File: spyder/plugins/editor/plugin.py [hunk 2]
```diff
@@ -327,10 +327,10 @@ def report_open_file(self, options):
...
                     self.completion_capabilities[language.lower()])
                 codeeditor.start_completion_services()
-            elif self.main.completions.is_fallback_only(language.lower()):
-                # This is required to use fallback completions for files
-                # without a language server.
-                codeeditor.start_completion_services()
+            # elif self.main.completions.is_fallback_only(language.lower()):
+            # This is required to use fallback completions for files
+            # without a language server.
+            codeeditor.start_completion_services()
```
Comment (ccordoba12) [inline spyder/plugins/editor/plugin.py:333]: I think this change is incorrect. We can't start the completion services unconditionally or things will break.

I mean, it took me a long time to fix this so I'd like to hear a good reason on why your changed this.

File: spyder/plugins/editor/widgets/editor.py [hunk 3]
```diff
@@ -621,7 +621,7 @@ def __init__(self, parent, actions):
         self.analysis_timer.timeout.connect(self.analyze_script)
 
         # Update filename label
-        self.editor_focus_changed.connect(self.update_fname_label)
+        self.editor_focus_changed.connect(lambda x, y: self.update_fname_label)
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/editor.py:624]:
```suggestion
        self.editor_focus_changed.connect(lambda x, y: self.update_fname_label())
```

I think this is required to call the corresponding method.

File: spyder/plugins/preferences/api.py [hunk 4]
```diff
@@ -37,6 +37,11 @@
 from spyder.widgets.comboboxes import FileComboBox
 
 
+class BaseConfigTab(QWidget):
+    """Stub class to declare a config page."""
```
Comment (ccordoba12) [inline spyder/plugins/preferences/api.py:41]:
```suggestion
    """Stub class to declare a config tab."""
```

File: spyder/plugins/pylint/plugin.py [hunk 5]
```diff
@@ -82,7 +82,8 @@ def register(self):
 
         # Connect to Editor
         widget.sig_edit_goto_requested.connect(editor.load)
-        editor.sig_editor_focus_changed.connect(self._set_filename)
+        editor.sig_editor_focus_changed.connect(
+            lambda x, y: self._set_filename)
```
Comment (ccordoba12) [inline spyder/plugins/pylint/plugin.py:86]:
```suggestion
            lambda x, y: self._set_filename())
```

I think this is also required (not really sure).

**Review by ccordoba12 — COMMENTED** [review]
Thanks a lot @andfoy for your hard work on this one! I left an initial review for you.

File: spyder/plugins/editor/plugin.py [hunk 6]
```diff
@@ -100,11 +100,25 @@ class Editor(SpyderPluginWidget):
 
     breakpoints_saved = Signal()
     run_in_current_extconsole = Signal(str, str, str, bool, bool)
-    open_file_update = Signal(str)
+
+    sig_file_opened_closed_updated = Signal(str, str)
```
Comment (ccordoba12) [inline spyder/plugins/editor/plugin.py:104]:
```suggestion
    sig_file_opened_closed_or_updated = Signal(str, str)
```

File: spyder/plugins/editor/plugin.py [hunk 7]
```diff
@@ -100,11 +100,25 @@ class Editor(SpyderPluginWidget):
...
+    language: str
+        Name of the programming language of the file that was opened,
+        closed or updated.
+    """
+
     sig_file_debug_message_requested = Signal()
 
     # This signal is fired for any focus change among all editor stacks
-    sig_editor_focus_changed = Signal()
+    sig_editor_focus_changed = Signal(str)
```
Comment (ccordoba12) [inline spyder/plugins/editor/plugin.py:121]:
```suggestion
    sig_editor_focus_changed = Signal()
```

File: spyder/plugins/editor/plugin.py [hunk 8]
```diff
@@ -1422,8 +1436,10 @@ def register_editorstack(self, editorstack):
                                               run_cell_copy))
         editorstack.update_plugin_title.connect(
                                    lambda: self.sig_update_plugin_title.emit())
-        editorstack.editor_focus_changed.connect(self.save_focus_editorstack)
-        editorstack.editor_focus_changed.connect(self.main.plugin_focus_changed)
+        editorstack.editor_focus_changed.connect(
+            lambda _x: self.save_focus_editorstack)
+        editorstack.editor_focus_changed.connect(
+            lambda _x: self.main.plugin_focus_changed)
```
Comment (ccordoba12) [inline spyder/plugins/editor/plugin.py:1442]: Revert

File: spyder/plugins/editor/widgets/editor.py [hunk 9]
```diff
@@ -423,7 +423,7 @@ class EditorStack(QWidget):
     run_cell_in_ipyclient = Signal(str, object, str, bool)
     debug_cell_in_ipyclient = Signal(str, object, str, bool)
     update_plugin_title = Signal()
-    editor_focus_changed = Signal()
+    editor_focus_changed = Signal(str)
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/editor.py:426]: Revert

File: spyder/plugins/editor/widgets/editor.py [hunk 10]
```diff
@@ -621,7 +621,7 @@ def __init__(self, parent, actions):
         self.analysis_timer.timeout.connect(self.analyze_script)
 
         # Update filename label
-        self.editor_focus_changed.connect(self.update_fname_label)
+        self.editor_focus_changed.connect(lambda x: self.update_fname_label)
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/editor.py:624]: Revert

File: spyder/plugins/editor/widgets/editor.py [hunk 11]
```diff
@@ -2401,14 +2405,16 @@ def tab_navigation_mru(self, forward=True):
...
         for finfo in self.data:
             if fwidget is finfo.editor:
+                fname = finfo.editor.filename
                 if finfo.editor.operation_in_progress:
                     self.spinner.start()
                 else:
                     self.spinner.stop()
                 self.refresh()
-        self.editor_focus_changed.emit()
+        self.editor_focus_changed.emit(fname)
```
Comment (ccordoba12) [inline spyder/plugins/editor/widgets/editor.py:2417]: Revert

File: spyder/plugins/pylint/plugin.py [hunk 12]
```diff
@@ -82,7 +82,8 @@ def register(self):
 
         # Connect to Editor
         widget.sig_edit_goto_requested.connect(editor.load)
-        editor.sig_editor_focus_changed.connect(self._set_filename)
+        editor.sig_editor_focus_changed.connect(
+            lambda x: self._set_filename)
```
Comment (ccordoba12) [inline spyder/plugins/pylint/plugin.py:86]: Revert

**Review by ccordoba12 — COMMENTED** [review]
@andfoy, this is a new review for you.

**Comment by dalthviz** [comment]
Regarding the Kite support I did an initial test but seems like enabling disabling kite is not being properly updated in the status bar:

After unchecking the enable kite config (and even after restarting Spyder) I still see the Kite logo and base text in the status bar:
image

Enabling/disabling Kite:

kite

**Comment by ccordoba12** [comment]
Hey @dalthviz, thanks for your feedback. I'm going to merge this PR now so we can use the observer pattern one (#14852) to finish the migration of the pending plugins. So please open a new issue about the problems you find with the Kite provider and @andfoy will take care of it in another PR.

**Review by ccordoba12 — APPROVED** [review]
Thanks a lot @andfoy for this!! It's a fantastic improvement!

**Comment by steff456** [comment]
In my testing of this PR all the options of the preference pane are working as expected! Thanks @andfoy for this work

**Pull request merged**